lib/ansible/modules/cloud/amazon/ec2_vpc_route_table_info.py

```diff
@@ -53,80 +60,174 @@
+              sample: active
+            vpc_peering_connection_id:
+              description: The ID of the VPC peering connection.
+              returned: when vpc peering available
+              type: str
+      tags:
+        description: Any tags assigned to the VPN connection.
+        returned: always
+        type: dict
+        sample: {
+                  "Name": "test-conn"
+                }
+      vpc_id:
+        description: The ID of the VPC peering connection.
+        returned: when vpc peering available
+        type: str
+        sample: vpc-a01106c2
+'''
 
+import json
 
-def get_route_table_info(route_table):
+try:
+    from botocore.exceptions import ClientError, BotoCoreError
+except ImportError:
+    pass  # caught by AnsibleAWSModule
 
-    # Add any routes to array
-    routes = []
-    associations = []
-    for route in route_table.routes:
-        routes.append(route.__dict__)
-    for association in route_table.associations:
-        associations.append(association.__dict__)
+from ansible.module_utils.aws.core import AnsibleAWSModule
+from ansible.module_utils.ec2 import (ansible_dict_to_boto3_filter_list, camel_dict_to_snake_dict,
+                                      boto3_tag_list_to_ansible_dict)
 
-    route_table_info = {'id': route_table.id,
-                        'routes': routes,
-                        'associations': associations,
-                        'tags': route_table.tags,
-                        'vpc_id': route_table.vpc_id
-                        }
 
-    return route_table_info
+def date_handler(obj):
+    return obj.isoformat() if hasattr(obj, 'isoformat') else obj
 
 
-def list_ec2_vpc_route_tables(connection, module):
+def list_route_tables(route_table, module):
+    params = dict()
 
-    filters = module.params.get("filters")
-    route_table_dict_array = []
+    params['Filters'] = ansible_dict_to_boto3_filter_list(module.params.get('filters'))
+    params['RouteTableIds'] = module.params.get('route_table_ids')
 
     try:
-        all_route_tables = connection.get_all_route_tables(filters=filters)
-    except BotoServerError as e:
-        module.fail_json(msg=e.message)
-
-    for route_table in all_route_tables:
-        route_table_dict_array.append(get_route_table_info(route_table))
-
-    module.exit_json(route_tables=route_table_dict_array)
+        result = json.loads(json.dumps(route_table.describe_route_tables(**params), default=date_handler))
+    except ValueError as e:
+        module.fail_json_aws(e, msg="Cannot validate JSON data")
+    except (ClientError, BotoCoreError) as e:
+        module.fail_json_aws(e, msg="Could not describe route tables")
+    snaked_route_tables = [camel_dict_to_snake_dict(route_table,
+                           ignore_list=['Tags']) for route_table in result['RouteTables']]
+    if snaked_route_tables:
+        for route_table in snaked_route_tables:
+            route_table['id'] = route_table['route_table_id']
```

Comment:
This looks like one you added for backward compatibility - do you mind also adding it to the return documentation? It would also be good to have a test for the return values from the boto version as a precaution - you could add something like this to the ec2_vpc_route_table integration test after https://github.com/ansible/ansible/blob/devel/test/integration/targets/ec2_vpc_route_table/tasks/main.yml#L493-L511:
```
- name: test route table facts
  ec2_vpc_route_table_info:
    filters:
      route-table-id: "{{ route_table_id }}"
    <<: *aws_connection_info
  register: result

- assert:
    that:
      - result.id
      - etc
```

Reply:
Sure. I will add it to the return documentation. I already added a test for checking if `id` is returned [here](https://github.com/ansible/ansible/pull/63413/files#diff-ea65d1c9ceeb31fc1b20355ec26f4dbfR593). I will add a test to check if `route_table_id` is also returned.
Is there any sort of deprecation policy to remove `id` from the returned values to stay in sync with boto3? Or are we now going to have it returned forever?

Reply:
The PR for `ec2_vpc_route_table_info` was merged earlier today - #63413

Reply:
Should I add the test case in a separate PR or should I rebase this on devel (so it includes #63413)?

Reply:
@prasadkatti Ideally the tests would be merged before porting to boto3, so if you don't mind making a second PR to get in first that would be awesome. I don't think a top level `id` is the only thing that changed between boto and boto3. Ideally there's be an assertion for each return value. I know that's tedious. `routes` and `associations` both seem to have more keys in the boto output here - https://github.com/ansible/ansible/pull/44966#issue-212382602 (that comment may be incomplete though, as there are no propagating_vgws - I don't know whether boto returned if they exist, and if so, whether those contents differ between boto and boto3 too).

We don't have a good way of deprecating return values at the moment, unfortunately. I left a comment here with some thoughts on that https://github.com/ansible/ansible/pull/62985#issuecomment-544986596. Feel free to weigh in if you have ideas!

Reply:
I have a question about this. The `route_tables.associations.id` (boto) has changed to `route_tables.associations.route_table_association_id` (boto3). Are we supposed to maintain backward compatibility for all such attributes? That sounds like a lot of work. I can try to do that if required. But can't we just add a release note for this module to let users know that some of the return values for this module have changed?

Reply:
We'll need a standard 4 release deprecation cycle to change the output, we always want to be sure we give people a fair chance to update their playbooks that may depend on return values or any other functionality. I haven't tested, but I would expect you should be able to set result.route_tables.associations.id to result.route_tables.associations.route_table_association_id before returning the result.  Then we would need the module RETURN docs and a porting guide entry to state that the return value is being deprecated.  #52518 has a good example of the needed docs.

Reply:
Thanks for the pointers. I am trying to gather a list of differences between boto and boto3 return values. This is what I have so far -

1. `route_tables.associations.id` is now `route_tables.associations.route_table_association_id`
2. Some fields seem to be returned conditionally in boto3. Before, in boto, they were always returned and would have a null value wherever applicable. Examples include `route_tables.associations.subnet_id`; `instance_id`, `interface_id` and `vpc_peering_connection_id` in `route_tables.routes`.
3. `propagating_vgws` - seems like a new return value

I know what needs to be done for (1). Don't think any action is needed for (3). For (2), do we want to add the missing keys with a null value?

Also, the above list is the difference that I could find. There might be more.

Reply:
I think it's better to set them to null if they always existed before - unknown key would behave differently than a key that exists but is set to a null value and we don't want to break things for people if possible. Thanks very much for diving in to this.

Reply:
@prasadkatti  At this point I think the remaining work is
a) set the values to None where boto would have done.  I don't particularly like this, but doing otherwise will break things for users.  The worst thing is we can't test for people trying to use it and spit out warnings.

b) Add some test cases for the return values.  test/integration/targets/ec2_vpc_route_table/tasks/main.yml already includes all of the framework so this should be relatively simple...